Fix log and partition replication documentation (#114)
restatedev/docs-restate: docs/server/clusters.mdx

- **Fast failover**: Followers can quickly take over as leader without reconstructing state from snapshots
- **Availability without durability overhead**: Additional partition replicas improve availability but do not increase storage costs, as durability comes from log replication and snapshots

**What it controls:**
- How many nodes simultaneously run processors for each partition
- Which node is the active leader processing invocations

### Log Replication

**Log replication** determines how many copies of the durable write-ahead log (Bifrost) are maintained across log-server nodes. This is the foundation of data durability in Restate.

**Key characteristics:**
- **Durability guarantee**: The log is the source of truth for all partition state changes
- **Quorum-based writes**: Log records are replicated to multiple nodes before being acknowledged
- **Survives node failures**: As long as a quorum of log-server nodes is available, the data is safe
- **Enables partition recovery**: Partitions can be restored on any node by replaying from the log or loading from snapshots

**What it controls:**

> - How many log-server nodes store copies of each log segment

Author: The more log-servers a record is stored on, the more log-server losses one can tolerate. However, it will also increase write amplification.

- The minimum number of nodes that must acknowledge a write

### Replication Properties

Both partition and log replication use the same **replication property** format, which can be either:

1. **Simple node count**: `2` means 2 replicas across any nodes
2. **Location-aware replication**: `{zone: 2, node: 3}` means 3 total replicas distributed across at least 2 zones

The `default-replication` setting in your configuration controls both partition and log replication by default, though they can be configured independently using `restatectl config set`.

### Location-Aware Replication

> Restate supports location-aware placement to improve fault tolerance across failure domains like zones and regions.

Author: Location-aware replication requires that node specify their location.

**Configuration:**
Each node can declare its location in the cluster configuration:

```toml restate.toml
# Specify the location of this node
location = "us-west.a1"  # format: "region[.zone]"
```

**Examples:**
- `"us-east"` — Region only
- `"eu-central.zone-1"` — Region and zone
- `""` (empty) — Default location (no specific assignment)

**Replication property examples:**
- `{node: 3}` — 3 replicas, distributed across any 3 nodes
- `{zone: 2, node: 3}` — 3 replicas, distributed across at least 2 different zones
- `{region: 2, zone: 3, node: 5}` — 5 replicas, across at least 2 regions and 3 zones

<Warning>
    Node locations should not change after initial node registration, especially for nodes running the `log-server` role, as this can affect replication guarantees.
</Warning>

### Relationship Between the Two

**Partition replication** and **log replication** work together to provide complete availability and durability:

1. **Log replication** ensures your data is durable and survives node failures
2. **Partition replication** ensures your application stays available by maintaining hot standby partition replicas

> 3. You can set higher partition replication for faster failover without increasing storage costs

Author: Higher partition replication increases storage and CPU costs as there are more nodes that actively apply the journal and store it in a materialized form in RocksDB.

4. You must set log replication high enough to tolerate node failures and maintain durability

**Example configuration:**
```toml restate.toml
# Replicate data to 3 nodes for durability
# Also run 3 partition replicas for fast failover
default-replication = 3
```

This cluster requires 3 nodes to become operational. With 5 nodes, it can tolerate 2 node failures while maintaining both availability (partition failover) and durability (log quorum).

> 

Author: We recommend to run with at most 2 partition processors unless you really have strict MTTR objectives.

**Key benefits:**
- **Fast failover**: Followers are already caught up with the log and can quickly take over as leader without reconstructing state from snapshots, reducing Mean Time To Recovery (MTTR) when a node fails

**Trade-offs:**
- Each partition replica consumes disk space and I/O resources to apply log records locally
- Does not improve data durability (that's the role of log replication)

**What it controls:**
- How many nodes simultaneously run partition processors that apply the log for each partition

### Log Replication

**Log replication** determines how many log-server nodes store copies of the durable write-ahead log (Bifrost). This is the foundation of data durability in Restate—the log stores the replicated data, not the partition processors.

**Key characteristics:**
- **Durability guarantee**: The log is the source of truth for all partition state changes. Your data is safe as long as you don't lose `n` log-server nodes, where `n` is the replication property value
- **Quorum-based writes**: Log records are replicated to multiple nodes before being acknowledged

> - **Enables partition recovery**: Partitions can be restored on any node by replaying from the log or loading from snapshots

Reviewer: Nit:

```suggestion
- **Enables partition recovery**: Partitions can be restored on any node by replaying the log or loading from snapshots
```

- `{zone: 2, node: 3}` — 3 replicas, distributed across at least 2 different zones
- `{region: 2, zone: 3, node: 5}` — 5 replicas, across at least 2 regions and 3 zones

<Warning>
    Node locations should not change after initial node registration, especially for nodes running the `log-server` role, as this can affect replication guarantees.
</Warning>

### Relationship Between the Two

**Partition replication** and **log replication** work together to provide complete availability and durability:

1. **Log replication** ensures your data is durable and survives node failures—the log stores the actual replicated data across log-server nodes
2. **Partition replication** ensures your application stays available by maintaining hot standby partition processors that apply the log
3. Partition processors don't store the replicated data; they apply log records to their local state. The log is the authoritative source
4. You must set log replication high enough to tolerate node failures and maintain durability

**Example configuration:**
```toml restate.toml
# Replicate log data to 2 nodes for durability
# Run 2 partition processors for fast failover
default-replication = 2
```

> This cluster requires 2 nodes to become operational. With 3 nodes, it can tolerate 1 node failure while maintaining both availability (partition failover) and durability (log quorum).

Author: ```suggestion
This cluster requires 2 nodes to become operational. With 3 nodes, it can tolerate 1 node failure while maintaining both availability of the system and durability of the data.
```

### Sizing Clusters

<Info>
    See [Understanding Partition and Log Replication](#understanding-partition-and-log-replication) for details on how replication works.
</Info>

The replicated log provider must be used with `default-provider = "replicated"` (this is the default).

> The `default-replication` determines the minimum number of nodes the data must be replicated to for both log and partition replication.

Author: The partition replication does not specify to how many nodes data must be replicated to but and how many partition processors are applying the log.
